src/EditorFeatures/Core/Implementation/Suggestions/SuggestedActionsSource.cs

```diff
@@ -860,6 +890,48 @@ private void OnSuggestedActionsChanged(Workspace currentWorkspace, DocumentId cu
 
                 Volatile.Write(ref _lastSolutionVersionReported, solutionVersion);
             }
+
+            public async Task<ISuggestedActionCategorySet> GetSuggestedActionCategoriesAsync(ISuggestedActionCategorySet requestedActionCategories, SnapshotSpan range, CancellationToken cancellationToken)
+            {
+                var provider = _owner;
+                var result = new List<string>();
+                using (var asyncToken = _owner.OperationListener.BeginAsyncOperation(nameof(GetSuggestedActionCategoriesAsync)))
+                {
+                    var selection = await GetSpanAsync(range).ConfigureAwait(false);
+                    if (selection != null)
+                    {
+                        var document = range.Snapshot.GetOpenDocumentInCurrentContextWithChanges();
+                        if (await HasFixesAsync(provider, document, range, cancellationToken).ConfigureAwait(false))
+                        {
+                            result.Add(PredefinedSuggestedActionCategoryNames.CodeFix);
+                            result.Add(PredefinedSuggestedActionCategoryNames.ErrorFix);
+                        }
+
+                        if (await HasRefactoringsAsync(provider, document, selection, cancellationToken).ConfigureAwait(false))
+                        {
+                            result.Add(PredefinedSuggestedActionCategoryNames.Refactoring);
+                        }
+                    }
+                }
+
+                return new SuggestedActionCategorySet(result);
```

Comment:
Creating your own implementation of this interface is fine, but if you have access to the ISuggestedActionCategoryRegistryService via MEF, you can use the [ISuggestedActionCategoryRegistryService.CreateSuggestedActionCategorySet()](https://msdn.microsoft.com/en-us/library/dn762984.aspx) method instead.